### PR TITLE
Icon Tooltip update for WCAG 2.1 compliance

### LINKED ISSUE #3094: Icon Tooltip: AVT 1 violation
## Environment

> Operating system

Mac OS X

> Browser

All

> Automated testing tool and ruleset

AAT v4.2.1

## Detailed description

Getting error from AAT test suite: `"wcag20.tech.g1204.inputVisibleLabel": "The input component is missing an associated visible label."`

> What version of the Carbon Design System are you using?

v10.3.1

> What did you expect to happen?

Not to get any AAT violations

> What happened instead?

Got a violation

## Steps to reproduce the issue

1. run the latest test suite of AAT on icon tooltip

## Additional information

This can be resolved by using a span within the button and making the text visually hidden in the default state and making it visible on focus/hover.  See Watson Health's demo for an example: https://pages.github.ibm.com/Watson-Health/components/detail/button--icon-only.html

### PR DESCRIPTION
**Pull request by elizabethsjudd**
Closes #3094 

Update to Icon Tooltip to pass AAT ruleset for WCAG 2.1

#### Changelog

**Changed**

- Icon Tooltip trigger now supports using a `span` within the button instead of the `aria-label` attribute.
- To support backwards compatibility and maintain a consistent animation, added `bx--tooltip--a11y` modifier to the trigger  button

#### Testing / Reviewing

- Should be tested in Carbon supported browsers (I believe it's Safari, Chrome, Firefox, Edge, and IE11)
- Use AAT (DAP) to verify the violations no longer are thrown

Before:
Screen Shot 2019-06-19 at 11 40 46 AM

After:
Screen Shot 2019-06-19 at 11 41 33 AM

### CODE REVIEW

**Review by emyarod — COMMENTED** [review]
my DAP results don't change from the WCAG 2.0 compliant tooltip to the WCAG 2.1 compliant tooltip, but I think it's because my DAP ruleset is out of date and I'm not sure how to upgrade

image

but from my other testing the WCAG 2.1 tooltip functions and appears like the WCAG 2.0 compliant tooltip

**Comment by elizabethsjudd** [comment]
@emyarod what version of DAP are you using?  The latest is `1.8.0.0` and the rulesets should match the screen shot below (make sure you are using the Internal IBM version of DAP that you download through this URL: https://aat.w3ibm.mybluemix.net/tools/auth/tools.html). The external version that you can access through the Chrome store is out of date and doesn't have any of the WCAG 2.1 updates. 

Screen Shot 2019-06-19 at 3 08 13 PM

**Comment by elizabethsjudd** [comment]
FYI, my plan was to remove the legacy versions from the demo (so they don't make it on to the website) once the legacy version has been verified to still work with the changes.

**Comment by joshblack** [comment]
@elizabethsjudd just curious, what were the violations that this PR is hoping to address?

File: packages/components/src/components/tooltip/tooltip--icon.hbs [hunk 1]
```diff
@@ -4,7 +4,81 @@
   This source code is licensed under the Apache-2.0 license found in the
   LICENSE file in the root directory of this source tree.
 -->
+<strong>WCAG 2.1 compliant</strong>
+<p>start</p>
+<br>
+{{!-- @todo Had to add a modifier to the parent to correct the animation but in the next major release it could be removed in favor of the new HTML format --}}
+<button
+  class="{{@root.prefix}}--tooltip__trigger {{@root.prefix}}--tooltip--a11y {{@root.prefix}}--tooltip__trigger--icon {{@root.prefix}}--tooltip--left {{@root.prefix}}--tooltip--align-start">
+  {{!-- @todo Open to suggestions for this element's className; In WH we have a utility class (wh-u-a11yTitle) that can be applied to other components --}}
```
Comment (scottnath) [inline packages/components/src/components/tooltip/tooltip--icon.hbs]: seems out-of-scope for this pr, but I'd vote for a global class for these a11y styles as they could be very useful in other situations. Then, `{{@root.prefix}}--tooltip__content` could bring those in for this use-case, but not have them locked to this component.

**Comment by elizabethsjudd** [comment]
@joshblack with WCAG 2.1 buttons that don't have text in them fail AAT because it reads it as missing a visible label.  By moving the content from `aria-label` to the span the tests are able to recognize that as the label.  WCAG 2.1 had a lot of updates where the use of `aria-label` is going to become a lot more restrictive.  Visuals and behaviors with this PR should remain the same as the previous version.

**Review by dakahn — APPROVED** [review]
This looks good to me. Thanks a ton @elizabethsjudd 🏄🏽‍♂️

I'd like @snidersd to get eyes on this since there's a fair few (seemingly much needed) changes. 👍🏽

File: packages/components/src/globals/scss/_tooltip.scss [hunk 2]
```diff
@@ -52,8 +55,26 @@
...
 
+  .#{$prefix}--tooltip__content {
+    width: 1px;
+    height: 1px;
+    padding: 0;
+    margin: -1px;
+    overflow: hidden;
+    clip: rect(0, 0, 0, 0);
+    border: 0;
+    visibility: visible;
```
Comment (asudoh) [inline packages/components/src/globals/scss/_tooltip.scss]: `visibility:visible` undoes `visiblity:hidden` in an ancestor node. I'd recommend `visibility:inherit`.

Comment (asudoh) [reply packages/components/src/globals/scss/_tooltip.scss]: Also does this have similar purpose of `.bx--assistive-text`? https://github.com/carbon-design-system/carbon/blob/v10.3.0/packages/components/src/globals/scss/_css--helpers.scss#L17-L29

**Review by asudoh — COMMENTED** [review]
Thanks @elizabethsjudd for jumping in!

@emyarod Would you have a chance to test it, especially considering long text (if you haven't done yet)? Thanks!

**Comment by emyarod** [comment]
@elizabethsjudd unfortunately I'm not able to view the page, I'm just taken to an  "Access Denied" even after authenticating. would it be possible to just share the .crx? not sure how to gain access to the a11y tools page you linked

@asudoh yeah it seems fine with long text and functions as expected, but I am seeing the tooltip caret finish animating before the tooltip body though (tagging @shixiedesign for design review). my guess is we can apply the same opacity + animation rules to the caret?

**Comment by elizabethsjudd** [comment]
@emyarod what is `the .crx`?  Also, try this link instead: http://ibm.biz/A11yToolsLandingPage (you have to be on the IBM VPN to access and use the internal DAP version)

**Comment by emyarod** [comment]
@elizabethsjudd the Chrome extension file (file extension .crx). I am accessing all of the links behind VPN but the latest link you provided times out during authentication for me

**Comment by elizabethsjudd** [comment]
@emyarod I found a link to the [chrome store](https://chrome.google.com/webstore/detail/dynamic-assessment-plugin/aahpafpbmmgednbflpalchnlbicaeppi) that has the correct version (1.8.0.0). If that still doesn't work I'll see if I can find the `.crx` file.  The fact that you can't get to the download links makes me a bit nervous though as those links are typically how you get your auth key that is required by the extension. 

Update: Tom Brunet posted in slack that there are some outages right now that are affecting the IMBa systems but that they should be back up now (11:10am EST)

**Comment by elizabethsjudd** [comment]
@emyarod @shixiedesign I noticed the difference in the animation as well.  The values for both animations are the same but one is using transition while the other is using keyframes so I could make that update to see if it fixes it.

Update: what was happening is that `transition` also applies when the user mouses off the trigger where keyframes don't so the animations were out of sync.  I don't know of a way (without using Javascript) to have a keyframe be triggered on "mouse out" so I applied the keyframe to both elements for the new version so the animations are at least consistent.

**Comment by emyarod** [comment]
yup I'm able to access the pages now @elizabethsjudd, I guess it was due to the outage. looks like my DAP extension was not auto updating even though I installed from the Chrome store previously

**Review by emyarod — COMMENTED** [review]
after retesting it looks like long content now flows out of the tooltip body

image

**Comment by elizabethsjudd** [comment]
@emyarod good catch!  I've made updates to support longer text (I've added the text to the demo temporarily for testing and will remove it before merging)

**Review by tw15egan — APPROVED** [review]

**Comment by asudoh** [comment]
@elizabethsjudd Thank you for the update - Another question I have is the reason behind going away from pseudo element. Do you have any...?

**Comment by elizabethsjudd** [comment]
@asudoh pseudo elements, that have actual content, are generally considered a bad practice in terms of a11y.  Some screen readers announce the content and some don't which causes an inconsistency, leading to either duplicate content or unaccessible content. Secondly, WCAG 2.1 had a primary focus on cognitive disabilities so newer guidelines are reducing the use of `aria-label` in favor of using `aria-labelledby` and referencing other content that is visible.  In the specific case of buttons the `aria-label` is not considered "visible text" even though Carbon had adapted it to be visible it wasn't picked up by automated testing (AVT 1: IBMa's AAT ruleset) to be "visible".  By using the span with in the button, automated testing can pick-up the content as "visible text".  This approach would also allow for the flexibility that if i user wanted to have a static text tooltip on a button that has content (think of the definition tooltip). The way that definition tooltip is currently implemented the user doesn't actually hear the content of the button, `aria-label` overwrites that content making it inaccessible for VoiceOver users.  I have another issue (#3099) where I plan to address that issue though.

**Comment by asudoh** [comment]
Thank you for explaining @elizabethsjudd. Several perspectives on this; We strive for simpler markup in our history, and avoid a11y concerns with contents in pseudo element by not putting the content in CSS but making the content reference to the markup one. Wrt our a11y tool; If it has a limitation, it should be fixed in the tool side. Though this specific instance is fine given definition tooltip used to have a similar markup, but just wanted to leave a note.

**Comment by elizabethsjudd** [comment]
@asudoh While I agree with you that simpler markup is better and work to follow this practice as well. Although I agree that it's a best practice to never define content in CSS this is typically a focus for i18n. In many cases this approach is not sufficient for a11y due to the latest changes to WCAG. WCAG 2.1 is a relatively new update (came out in April/May i believe) but IBMa has has released updates available to support the new guidelines.  IBMa's tools are the most robust automation testing we have available and we should be following their latest ruleset with all of our testing.

File: packages/components/src/globals/scss/_tooltip.scss [hunk 3]
```diff
@@ -31,16 +31,23 @@
     }
   }
 
-  // Tooltip - renders as a combo of ::before and ::after elements
+  // ::before - Tooltip caret
+  // ::after - legacy Tooltip content (depreciated)
```
Comment (emyarod) [inline packages/components/src/globals/scss/_tooltip.scss]:
```suggestion
  // ::after - legacy Tooltip content (deprecated)
```

**Review by emyarod — APPROVED** [review]
looks good to me, thanks for the quick fix for long text!

**Comment by joshblack** [comment]
@elizabethsjudd for:

> @joshblack with WCAG 2.1 buttons that don't have text in them fail AAT because it reads it as missing a visible label. By moving the content from aria-label to the span the tests are able to recognize that as the label. WCAG 2.1 had a lot of updates where the use of aria-label is going to become a lot more restrictive. Visuals and behaviors with this PR should remain the same as the previous version

Could you point me to the new guidelines so I can study up? 🙏

**Review by joshblack — APPROVED** [review]

**Comment by joshblack** [comment]
cc @shixiedesign for review

**Comment by elizabethsjudd** [comment]
@joshblack Here's the official docs: https://www.w3.org/WAI/standards-guidelines/wcag/new-in-21/

I've honestly gotten most of my information on updates to 2.1 from webinars both by IBMa and deque (creators of aXe).  IBMa has monthly calls that go over "what's new" in a11y which is typically my starting point to keep on new things.  I found out about the deque meeting from an IBMa meeting.

**Comment by joshblack** [comment]
Thanks so much @elizabethsjudd! Really appreciate it 🙏

**Review by shixiedesign — COMMENTED** [review]
Thank you! And especially thanks for double checking the motion part too 🙏 

After finally getting to do the review, I noticed a tiny gap between the triangle of the tooltip & the tooltip body box. Vanilla only. @emyarod explained this is also magnification dependent so I'm not sure if there's a fix. If not I guess it's not a big deal, lemme know @elizabethsjudd Everything else looks great! 

image

**Comment by elizabethsjudd** [comment]
@shixiedesign thanks for the review!  I've noticed the gap between the body and caret as well but I saw it happen with the current implementation.  As I was working on the definition tooltip, I noticed that the caret placement was actually quite finicky based on surrounding components/styles that were having odd side-effects with the component. I do feel like it's a valid issue to re-access the caret placement given all of the available demos, but that it's beyond the scope of this PR where I'm really focused on the a11y of the component while maintaining existing behaviors and styles.

**Comment by asudoh** [comment]
@elizabethsjudd Though we don't have a guideline for this topic, it'd be great if you can figure out a way to avoid `:not()` - The way I tend to avoid `:not()` is having some overriding style that negates original. Probably `.bx--tooltip--a11y::after() { content: none; }` (more specificity may be needed) for your case?

**Review by shixiedesign — APPROVED** [review]
👍 agreed! Thanks for the fix!

**Comment by elizabethsjudd** [comment]
@asudoh out of curiosity, why do you avoid the `:not` selectors (i've heard pros and cons both ways and was just curious for your reasoning)

**Review by emyarod — COMMENTED** [review]
meant to comment in #3151

**Comment by tw15egan** [comment]
@asudoh @elizabethsjudd I'm also curious as to why we should avoid `:not` selectors. I thought the optimization of `css` selectors was pretty negligible.

**Comment by asudoh** [comment]
@elizabethsjudd @tw15egan Good question from you both! The event I lead to my inclination to avoid `:not()` has completely slipped off my mind, but a couple of things what I can recall... One is every occurrence of "I don't want this style for this feature" ends up with addition to `:not()`, like `:not(.bx--foo):not(.bx--bar):not(.bx--baz)`. Also, the style rules of a feature set ends up with with multiple places, one is where `:not()` is, another is the feature-specific style implementation itself. I'm inclined to having `.bx--foo { ... }` work for `:not()` portion as well as for others.

**Comment by elizabethsjudd** [comment]
@asudoh I agree with the fact that `:not` should be used sparingly in the fact that yes, if you just use it as a fail safe you do end up with bloated selectors as you described. Another reason I have found the `:not` to be a pain is that it tends leads to higher selector weight making them harder to overwrite. However, in cases like this where we are trying to maintain backwards compatibility I do feel like they are a wash.  We still end up with multiple feature specific declarations (the default backwards compatible and then overwriting for the new feature). Right now in tooltip the `after` pseudo element is actually still getting all of the styles applied to it even though it doesn't use them.  which could actually lead to unexpected behaviors for users of Carbon that are trying to extend the styles.  I'm good with keeping them removed but I guess I'm saying we shouldn't treat it as a blanket "never use it" but instead assess the situation and figure what is the appropriate solution.  It was added to the CSS specification for a reason.

**Comment by asudoh** [comment]
Thank you for sharing your thoughts @elizabethsjudd - If I can focus on just one thing, you are right that `::after()` gets all the styles even though they are not in use. As far as I see `content:none` eliminates all the adverse effects by that. And your latest code with `content:none` seems to have made the change look more straightforward.

**Comment by tw15egan** [comment]
@asudoh thanks for filling me in! Makes sense, and I agree in this situation. @elizabethsjudd hit it on the head here:

>I guess I'm saying we shouldn't treat it as a blanket "never use it" but instead assess the situation and figure what is the appropriate solution

**Comment by asudoh** [comment]
@tw15egan Yes makes sense! - What I meant by my first comment wrt `:not()` is that good to think twice when we are tempted to use `:not()`.

**Comment by asudoh** [comment]
@dakahn Do you think we are good to go? Or should we wait @snidersd? Thanks!

**Comment by dakahn** [comment]
@asudoh I'm happy! Let's do it 🏄

**Pull request merged**